doc/source/reference/c-api.dtype.rst

```diff
@@ -374,3 +374,22 @@ format specifier in printf and related commands.
     :c:data:`NPY_LONGLONG_FMT`, :c:data:`NPY_ULONGLONG_FMT`,
     :c:data:`NPY_INTP_FMT`, :c:data:`NPY_UINTP_FMT`,
     :c:data:`NPY_LONGDOUBLE_FMT`
+
+Generic Typing and Deviations from C Standard Syntax
+----------------------------------------------------
+
+Many of the C source files (extension: :file:`.c.src`) in NumPy are not
+compliant with standard C syntax because the :c:data:`@` symbol is used
+as a substitution mechansim for the large number of data types described above.
+For example, a loop in the NumPy C source code may have a :c:data:`@TYPE@`
+variable which is preprocessed to a number of otherwise identical loops
+with several types such as :c:data:`INT, LONG, UINT, ULONG`.
+The :c:data:`@TYPE@` style syntax thus reduces code duplication and
+maintenance burden by mimicking languages that have generic type support.
+By convention, generically typed blocks are preceded by comment blocks that
+enumerate the intended type substitutions.
+
+The preprocessing of generically typed NumPy C source files is performed
+by a set of Python modules (:file:`numpy/core/code_generators`). The type
```

Comment:
The macros you're trying to document aren't expanded by anything in the numpy/core/code_generators directory.  The code you're looking for is part of numpy-distutils in numpy/distutils/conv_template.py.  

I think the "C standard compliant" wording here is a little negative about this.  Numpy uses a custom code generator that it also ships.  This isn't really all that crazy of a thing to do.  Its not like doing this with C macros or just accepting that massive code duplication are either good choices.

Reply:
Interesting, and I see there are more "rules" than I thought in `numpy/distutils/conv_template.py`. I'll try to rephrase to make it a bit less framed around C standard compliance deviations, though this should likely be mentioned at least somewhere for clarity.